File: src/dotnet/PolyBus/Transport/Transactions/Messages/Handlers/Serializers/JsonHandlers.cs
using System.Text.Json;

namespace PolyBus.Transport.Transactions.Messages.Handlers.Serializers;

/// <summary>
/// Handlers for serializing and deserializing messages as JSON.
/// </summary>
public class JsonHandlers
{
    /// <summary>
    /// The options to use for the JSON serializer.
    /// </summary>
    public JsonSerializerOptions? JsonSerializerOptions { get; set; }

    /// <summary>
    /// The content type to set on outgoing messages.
    /// </summary>
    public string ContentType { get; set; } = "application/json";

    /// <summary>
    /// The header key to use for the content type.
    /// </summary>
    public string Header { get; set; } = Headers.ContentType;

    /// <summary>
    /// Deserializes incoming messages from JSON.
    /// </summary>
    public virtual Task Deserializer(IncomingTransaction transaction, Func<Task> next)
    {
        var incomingMessage = transaction.IncomingMessage;

        incomingMessage.Message = JsonSerializer.Deserialize(incomingMessage.Body, incomingMessage.MessageType, JsonSerializerOptions)!;

        return next();
    }

    /// <summary>
    /// Serializes outgoing messages to JSON.
    /// </summary>
    public virtual Task Serializer(OutgoingTransaction transaction, Func<Task> next)
    {
        foreach (var message in transaction.OutgoingMessages)
        {
            message.Body = JsonSerializer.Serialize(message.Message, JsonSerializerOptions);
            message.Headers[Header] = ContentType;
        }
        return next();
    }
}

File: src/kotlin/src/main/kotlin/polybus/transport/transactions/messages/handlers/serializers/JsonHandlers.kt
package polybus.transport.transactions.messages.handlers.serializers

import polybus.Headers
import polybus.transport.transactions.IncomingTransaction
import polybus.transport.transactions.OutgoingTransaction
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule

open class JsonHandlers {
    open var objectMapper: ObjectMapper = ObjectMapper().registerKotlinModule()
    open var contentType: String = "application/json"
    open var header: String = Headers.CONTENT_TYPE

    open suspend fun deserializer(transaction: IncomingTransaction, next: suspend () -> Unit) {
        val incomingMessage = transaction.incomingMessage
        incomingMessage.message = objectMapper.readValue(incomingMessage.body, incomingMessage.messageType)
        next()
    }

    open suspend fun serializer(transaction: OutgoingTransaction, next: suspend () -> Unit) {
        transaction.outgoingMessages.forEach { message ->
            message.body = objectMapper.writeValueAsString(message.message)
            message.headers[header] = contentType
        }
        next()
    }
}
